qutebrowser/misc/notree.py

```diff
@@ -0,0 +1,240 @@
+
+
+class TraverseOrder(enum.Enum):
+    """To be used as argument to traverse().
+
+    Implemented orders are pre-order and post-order.
+    Attributes:
+        PRE: pre-order (parents before children). Same as in Node.render
+        POST: children of a node are always yield before their parent.
+    """
+
+    PRE = enum.auto()
+    POST = enum.auto()
+
+
+class Node():
+    """Fundamental unit of notree library.
+
+    Attributes:
+        value: The element (ususally a tab) the node represents
+        parent: Node's parent.
+        children: Node's children elements.
+        siblings: Children of parent node that are not self.
+        path: List of nodes from root of tree to self
+    value, parent, and children can all be set by user. Everything else will be
+    updated accordingly, so that if `node.parent = root_node`, then `node in
+    root_node.children` will be True.
+    """
+
+    sep = '/'
+    __parent: 'Node' = None
+
+    def __init__(self, value, parent=None, childs=()):
+        self.value = value
+        # set initial values so there's no need for AttributeError checks
+        self.__parent = None
+        self.__children = []
+
+        # For render memoization
+        self.__modified = False
+        self.__set_modified()  # not the same as line above
+        self.__rendered = []
+
+        if parent:
+            self.parent = parent  # calls setter
+        if childs:
+            self.children = childs  # this too
+
+        self.collapsed = False
+
+    @property
+    def parent(self):
+        return self.__parent
+
+    @parent.setter
+    def parent(self, value):
+        """Set parent property. Also adds self to value.children."""
+        # pylint: disable=protected-access
+        assert(value is None or isinstance(value, Node))
+        if self.__parent:
+            self.__parent.__set_modified()
+            self.__parent.__disown(self)
+            self.__parent = None
+        if value is not None:
+            value.__add_child(self)
+            self.__parent = value
+        self.__set_modified()
+
+    @property
+    def children(self):
+        return tuple(self.__children)
+
+    @children.setter
+    def children(self, value):
+        """Set children property, preserving order.
+
+        Also sets n.parent = self for n in value. Does not allow duplicates.
+        """
+        seen = set(value)
+        if len(seen) != len(value):
+            raise TreeError("A duplicate item is present in in %r" % value)
+        new_children = list(value)
+        for child in new_children:
+            if child.parent is not self:
+                child.parent = self
+        self.__children = new_children
+        self.__set_modified()
+
+    @property
+    def path(self):
+        """Get a list of all nodes from the root node to self."""
+        if self.parent is None:
+            return [self]
+        else:
+            return self.parent.path + [self]
```

Comment:
We could also just use a set, just like in `children.setter`. The `notree` library does something similar to avoid loops; 
Right now the browser crashes if there's any loops and `path` is accessed. I think it's not a bad idea to let it crash; the problem wouldn't be here as much as someplace else, because there should never ever be loops. The most I'd do is to add a check and then still raise a (more descriptive) exception.